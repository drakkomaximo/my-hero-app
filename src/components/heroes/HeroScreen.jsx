import React, { useMemo } from "react";
import { Redirect, useParams } from "react-router";
import { getHeroesById } from "../selectors/getHeroesById";

const HeroScreen = ({ history }) => {
  const { heroId } = useParams();

  const hero = useMemo(() => getHeroesById(heroId), [ heroId ])

  if (!hero) {
    return <Redirect to="/" />;
  }

  const {
    superhero,
    publisher,
    alter_ego,
    first_appearance,
    characters,
    url_main
  } = hero;

  const handleReturn = () => {

    history.replace(`/${url_main}`)

  };

  return (
    <div className="row mt-3">
      <div className="col-4">
        <img
          src={`../assets/heroes/${heroId}.jpg`}
          alt={superhero}
          className="img-thumbnail animate__animated animate__fadeInLeft"
        />
      </div>
      <div className="col-8 animate__animated animate__fadeInRight">
        <h3>{superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            <b>Alter ego:</b> {alter_ego}
          </li>
          <li className="list-group-item">
            <b>Publisher:</b> {publisher}
          </li>
          <li className="list-group-item">
            <b>First Apperance:</b> {first_appearance}
          </li>
        </ul>

        <h5> Characters </h5>
        <p> {characters}</p>

        <button className="btn btn-outline-info" onClick={handleReturn}>
          Return
        </button>
      </div>
    </div>
  );
};

export default HeroScreen;
