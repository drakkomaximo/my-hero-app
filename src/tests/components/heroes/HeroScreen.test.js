import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router"
import HeroScreen from "../../../components/heroes/HeroScreen"

describe('Tests on <HeroScreen />', () => {
    
    const historyMock = {
        replace: jest.fn(),
    }
    
    test('It must show correctly', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero']}>
                <HeroScreen history={ historyMock }/>
            </MemoryRouter>
        )

        expect( wrapper ).toMatchSnapshot()
    })

    test('It must show redirect component if there are not arguments in the URL', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero']}>
                <HeroScreen history={ historyMock }/>
            </MemoryRouter>
        )

        expect( wrapper.find('Redirect').exists() ).toBe(true)
    })

    test('It must show a hero if the parameter exists and finds it', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/bha-dynamight']}>
                <Route path='/hero/:heroId' component={ HeroScreen }/>
            </MemoryRouter>
        )

        expect( wrapper.find('.row').exists() ).toBe(true)
    })

    test('It must call the HandleReturn function and replace function should have been called', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/bha-dynamight']}>
                <Route 
                    path='/hero/:heroId' 
                    component={ () => <HeroScreen history={historyMock}/>  }/>
            </MemoryRouter>
        )

        wrapper.find('button').prop('onClick')()
        expect( historyMock.replace ).toHaveBeenCalledTimes(1)
        expect( historyMock.replace ).toHaveBeenCalledWith('/bha')

    })

    test('It must call redirect if the hero does not exists', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/bha-dynamightdfsfdsdf']}>
                <Route 
                    path='/hero/:heroId' 
                    component={ () => <HeroScreen history={historyMock}/>  }/>
            </MemoryRouter>
        )

        expect( wrapper.text() ).toBe('')
    })
    
    
    
    
})
