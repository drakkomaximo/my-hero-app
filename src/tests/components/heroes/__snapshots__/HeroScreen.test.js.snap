// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <HeroScreen /> It must show correctly 1`] = `""`;
